File: 2-AddTwoNumbers/2-AddTwoNumbers.cpp
// Last updated: 11/7/2025, 2:49:49 PM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* res = new ListNode(); // Инициализируем указатель на новый узел
        ListNode* current1 = l1;
        ListNode* current2 = l2;
        ListNode* currentRes = res; // Используем отдельный указатель для результата

        int carry = 0; // Переменная для хранения переноса

        while (current1 != nullptr || current2 != nullptr) {
            int x = (current1 != nullptr) ? current1->val : 0;
            int y = (current2 != nullptr) ? current2->val : 0;
            int sum = x + y + carry;
            carry = sum / 10;
            currentRes->next = new ListNode(sum % 10);
            if (current1 != nullptr) current1 = current1->next;
            if (current2 != nullptr) current2 = current2->next;
            currentRes = currentRes->next;
        }
        if (carry > 0) {
            currentRes->next = new ListNode(carry);
        }
        return res->next; 
    }
};

File: 2-AddTwoNumbers/2-AddTwoNumbers.go
// Last updated: 11/7/2025, 2:49:47 PM
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 // 18 % 10 = 8
 // 18 // 10 = 1
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    curr1, curr2 := l1, l2
    dummy := &ListNode{}
    result := dummy
    var carry int
    for curr1 != nil || curr2 != nil || carry != 0 {
        curr1Val, curr2Val := 0, 0
        if curr1 != nil {
            curr1Val = curr1.Val
            curr1 = curr1.Next
        }
        if curr2 != nil {
            curr2Val = curr2.Val
            curr2 = curr2.Next
        }
        
        sum := curr1Val + curr2Val + carry
        carry = sum / 10
        digit := sum % 10

        result.Next = &ListNode{Val: digit}

        result = result.Next
    }

    return dummy.Next
}